src/Microsoft.AspNetCore.OData/Batch/ODataBatchReaderExtensions.cs

```diff
@@ -140,9 +144,21 @@ public static class ODataBatchReaderExtensions
 
             foreach (var header in batchRequest.Headers)
             {
-                // Copy headers from batch, overwriting any existing headers.
                 string headerName = header.Key;
                 string headerValue = header.Value;
+
+                if (headerName.Trim().ToLowerInvariant() == "prefer")
+                {
+                    // in the case of Prefer header, we don't want to overwrite,
+                    // instead we merge preferences defined in the individual request with those inherited from the batch
+                    request.Headers.TryGetValue(headerName, out StringValues batchPreferencesList);
+                    // TODO: what if multiple Prefer headers are defined on the request?
```

Comment:
>what if multiple Prefer headers are defined on the request [](start = 29, length = 58)

proper behavior would be to merge them, giving preference to the first one specified

From [https://tools.ietf.org/html/rfc7240](https://tools.ietf.org/html/rfc7240) :
" To avoid any possible ambiguity, individual preference tokens SHOULD
   NOT appear multiple times within a single request.  If any preference
   is specified more than once, only the first instance is to be
   considered.  All subsequent occurrences SHOULD be ignored without
   signaling an error or otherwise altering the processing of the
   request.  This is the only case in which the ordering of preferences
   within a request is considered to be significant."

Reply:
I found that the batch reader throws an exception if an individual request includes a duplicate header (e.g. if it assigns the Prefer header twice), so that made things easier.

But I have accounted for the case where the batch requests happens to have multiple occurrences of the Prefer header and multiple occurrences of individual preferences. The `StringValues` returned when you access a header in `request.Headers` can already correctly combine the values from multiple occurrences of the header into one string. So all I did was just make sure that duplicate occurrences of individual preferences are ignored.